read-fonts/src/lib.rs

```diff
@@ -71,6 +71,86 @@ pub(crate) mod codegen_prelude {
 
 include!("../generated/font.rs");
 
+#[derive(Clone)]
+/// Reference to the content of a font or font collection file.
+pub enum FileRef<'a> {
+    /// A single font.
+    Font(FontRef<'a>),
+    /// A collection of fonts.
+    Collection(CollectionRef<'a>),
+}
+
+impl<'a> FileRef<'a> {
+    pub fn new(data: FontData<'a>) -> Result<Self, ReadError> {
+        Ok(if let Ok(collection) = CollectionRef::new(data) {
+            Self::Collection(collection)
+        } else {
+            Self::Font(FontRef::new(data)?)
+        })
+    }
+
+    pub fn fonts(&self) -> impl Iterator<Item = Result<FontRef<'a>, ReadError>> + 'a + Clone {
+        let copy = self.clone();
+        let len = match self {
+            Self::Collection(collection) => collection.len(),
+            _ => 1,
+        };
+        (0..len).filter_map(move |ix| match &copy {
+            Self::Font(font) => Some(Ok(font.clone())),
+            Self::Collection(collection) => collection.get(ix),
+        })
+    }
+}
+
+const TTC_HEADER_TAG: Tag = Tag::new(b"ttcf");
+
+/// Reference to the content of a font collection file.
+#[derive(Clone)]
+pub struct CollectionRef<'a> {
+    data: FontData<'a>,
+    header: TTCHeader<'a>,
+}
+
+impl<'a> CollectionRef<'a> {
+    pub fn new(data: FontData<'a>) -> Result<Self, ReadError> {
+        let header = TTCHeader::read(data)?;
+        if header.ttc_tag() != TTC_HEADER_TAG {
+            Err(ReadError::InvalidTtc(header.ttc_tag()))
+        } else {
+            Ok(Self { data, header })
+        }
+    }
+
+    pub fn len(&self) -> u32 {
+        self.header.num_fonts()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn get(&self, index: u32) -> Option<Result<FontRef<'a>, ReadError>> {
+        if index >= self.len() {
+            return None;
+        }
+        let offset = self
+            .header
+            .table_directory_offsets()
+            .get(index as usize)?
+            .get() as usize;
+        match TableDirectory::read(self.data.slice(offset..)?) {
+            Ok(table_directory) => Some(FontRef::with_table_directory(self.data, table_directory)),
+            _ => None,
+        }
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = Result<FontRef<'a>, ReadError>> + 'a + Clone {
```

Comment:
again don't think we need `Clone`?

And also for fun (it's fine either way) the alternative to the inner clone would be,

```suggestion
    pub fn iter<'b: 'a>(&'b self) -> impl Iterator<Item = Result<FontRef<'a>, ReadError>> + 'a {
```

although I honestly might prefer your version just for avoiding the confusion of an extra lifetime in the signature.

Reply:
Yeah, adding a lifetime to `&self` can be annoying for many reasons and often breaks method chaining forcing you to store temporaries in a variable